src/machine/image.c

```diff
@@ -0,0 +1,205 @@
+                return log_debug_errno(result_fd, "Failed to open tmpfile: %m");
+
+        /* This might be a slow operation, run it asynchronously in a background process */
+        r = safe_fork("(sd-clean)", FORK_RESET_SIGNALS, &child);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to fork(): %m");
+        if (r == 0) {
+                _cleanup_hashmap_free_ Hashmap *images = NULL;
+                bool success = true;
+                Image *image;
+
+                errno_pipe_fd[0] = safe_close(errno_pipe_fd[0]);
+
+                images = hashmap_new(&image_hash_ops);
+                if (!images)
+                        report_errno_and_exit(errno_pipe_fd[1], EIO);
+
+                r = image_discover(manager->runtime_scope, IMAGE_MACHINE, /* root = */ NULL, images);
+                if (r < 0) {
+                        log_debug_errno(r, "Failed to discover images: %m");
+                        report_errno_and_exit(errno_pipe_fd[1], r);
+                }
+
+                ssize_t n = loop_write(result_fd, &success, sizeof(success));
+                if (n < 0) {
+                        log_debug_errno(n, "Failed to write to tmp file: %m");
+                        report_errno_and_exit(errno_pipe_fd[1], n);
+                }
+
+                HASHMAP_FOREACH(image, images) {
+                        /* We can't remove vendor images (i.e. those in /usr) */
+                        if (IMAGE_IS_VENDOR(image))
+                                continue;
+                        if (IMAGE_IS_HOST(image))
+                                continue;
+                        if (mode == IMAGE_CLEAN_POOL_REMOVE_HIDDEN && !IMAGE_IS_HIDDEN(image))
+                                continue;
+
+                        r = image_remove(image);
+                        if (r == -EBUSY) {
+                                log_debug("Keep image '%s' because it's currently used.", image->name);
+                                continue;
+                        }
+                        if (r < 0) {
+                                log_debug_errno(r, "Failed to remove image '%s': %m", image->name);
+                                success = false;
+
+                                /* If the operation failed, let's override everything we wrote, and instead write there at which image we failed. */
+                                int k = ftruncate(result_fd, 0);
+                                if (k >= 0)
+                                        k = lseek(result_fd, 0, SEEK_SET);
+                                if (k >= 0)
+                                        k = write(result_fd, &success, sizeof(success));
+                                if (k >= 0)
+                                        k = write(result_fd, image->name, strlen(image->name)+1);
+                                if (k < 0)
+                                        log_debug_errno(k, "Failed to truncate, rewind, or write to tmp file: %m");
```

Comment:
i think this would be a lot more readable if these 4 calls would be a function on their own.

there's some type confusion here. lseek() returns off_t, write() returns ssize_t. and ftruncare() returns int. on 64bit off_t + ssize_t are 64bit, and thus the values don't fit in `k`. on 32bit off_t is still 64bit, and thus things are broken too.

please use variables with the right type, do not accidently drop half the bits.

Reply:
I'll fix the type problem. Thanks for spotting it. But I'd suggest to keep code as is. A separate function would be too specific and unlikely for reuse.